sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -88,16 +88,27 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
         s"Selected $selected partitions out of $total, pruned $percentPruned% partitions."
       }
 
+      // need to add projections from combineFilters in
+      val combineFilter = combineFilters.reduceLeftOption(expressions.And)
+      val combinedProjects = combineFilter.map(_.references.toSet.union(projects.toSet).toSeq)
+        .getOrElse(projects)
       val scan = buildPartitionedTableScan(
         l,
-        projects,
+        combinedProjects,
         pushedFilters,
         t.partitionSpec.partitionColumns,
         selectedPartitions)
 
-      combineFilters
-        .reduceLeftOption(expressions.And)
-        .map(execution.Filter(_, scan)).getOrElse(scan) :: Nil
+      // Add a Projection to guarantee the original projection:
+      // this is because "combinedProjects" may be different from the
+      // original "projects", in elements or their ordering
```

Comment:
Thanks for the suggestion. Thought the == is 'shallow' compare, but it is 'deep' equality checking. Will make the changes.